### PR TITLE
Introduce a common parent interface for JedisClusterCommands and JedisCommands

### PR DESCRIPTION
**Pull request by ptahchiev**
Unify the common methods from JedisClusterCommands and JedisCommands in a common parent interface to avoid duplication.

### CODE REVIEW

File: src/main/java/redis/clients/jedis/commands/JCommands.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Petar Tahchiev
+ * @since 2.0.1
+ */
+public interface JCommands {
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/commands/JCommands.java]: I think we need a better naming

**Comment by gkorland** [comment]
in general it LGTM, but
1. I think we need a better name than JCommands
2. @marcosnils was there any logic behind this duplication ?

File: src/main/java/redis/clients/jedis/commands/JCommands.java (hunk 1 above)
Comment (ptahchiev) [reply src/main/java/redis/clients/jedis/commands/JCommands.java]: Agree

Comment (ptahchiev) [reply src/main/java/redis/clients/jedis/commands/JCommands.java]: I have renamed `JCommands` to `JedisBaseCommands`

**Comment by gkorland** [comment]
LGTM!

**Comment by marcosnils** [comment]
LGTM!. I've assigned to 4.0.0 as it breaks backwards compatibility.

**Comment by gkorland** [comment]
@marcosnils  how does it break?

**Comment by marcosnils** [comment]
@gkorland for projects that were using JedisCommands or JedisClusterCommands the code won't compile anymore because methods were moved to a new interface. We're breaking the API for external consumers

**Comment by gkorland** [comment]
@marcosnils But that is the all idea, JedisCommands and JedisClusterCommands interfaces extend the new JedisBaseCommands interface, is should still compile.

**Comment by ptahchiev** [comment]
@marcosnils why would it not compile when the existing interfaces extend from the new interface?

**Comment by ptahchiev** [comment]
This PR is marked as ready to merge, but since 2 months nothing has happened. And now it is unmergeable, because both interfaces added new methods. And guys what: `JedisClusterCommands` methods use `Long` while `JedisCommands` use `long`. This wouldn't have happened if both methods were in the same parent interface like I suggested. 
@gkorland @marcosnils  please provide some constructive feedback on why you don't want this merged or how to improve it so I can get it merged.

**Comment by gkorland** [comment]
@ptahchiev you're right, I don't see a reason not to merge it to 3.1, but @marcosnils had some concerns.
BTW, I tried to merge/fix the branch using GH merge GUI but it seems like I broke your branch, can you please review it.

**Comment by ptahchiev** [comment]
@gkorland Like I said now some of the methods have diverged. Two methods have the same name and parameters, but one returns `java.lang.Long` and the other one returns `long`. I can rewrite my PR but now it will  bigger. Can you please advise me which return signature to use - `java.lang.Long` or `long` is correct?

**Comment by sazzad16** [comment]
@gkorland I do not support to merge this in 3.x.

If you really want to merge this now, I'd suggest to merge this ONLY in master (which will act as 4.0). Then we'd have to manage 4.0/master and 3.x separately.

File: src/main/java/redis/clients/jedis/commands/JedisBaseCommands.java [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+import redis.clients.jedis.params.SetParams;
+import redis.clients.jedis.params.ZAddParams;
+import redis.clients.jedis.params.ZIncrByParams;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Petar Tahchiev
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/JedisBaseCommands.java]: Please remove author tag.

File: src/main/java/redis/clients/jedis/commands/JedisBaseCommands.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+ * @since 2.0.1
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/JedisBaseCommands.java]: ???

**Comment by sazzad16** [comment]
@ptahchiev @gkorland @marcosnils 

I'm a bit concerned about the name `JedisBaseCommands`. What are trying to mean by the word `Base`? Furthermore, this may create confusion between `JedisBaseCommands` and `BasicCommands` which contain separate methods.

**So, I would strongly suggest a name change.**

Just dropping a suggestion: may be something like, `JedisCommonCommands`?

**Comment by ptahchiev** [comment]
@sazzad16 I have no problem with a name change. However, I do believe that this should be merged in the 3.x branch as it is not a breaking change and is in fact a blocker to other open-source projects (jedis-quartz, spring-boot with jedis-quartz, etc). @sazzad16  is the name change your only objection?

**Comment by sazzad16** [comment]
@ptahchiev You've asked a simple but complicated question! Before answering anything, I purposefully avoided myself from participating here. because I don't exactly support this PR with the idea of introducing a new common parent interface.

I lean more towards **unifying** the concerned interfaces. This idea was expressed by @HeartSaVioR earlier in #1545.<sup>[1]</sup>

[1]: https://github.com/xetorthio/jedis/issues/1545#issuecomment-347687167

**Comment by ptahchiev** [comment]
Hi @sazzad16 I am totally OK with unifying both the interfaces into one. Like I said I raised this issue (https://github.com/xetorthio/jedis/issues/1931) because I cannot upgrade the quartz-redis-jobstore to the jedis 3.0 and I am really blocked. I would really like to see a decision being made and some progress on this issue.

**Comment by sazzad16** [comment]
@ptahchiev Thank you for understanding!

And, for this approach, you've already got two collaborators participating which is enough! I just spoke out when I felt to, for the format of the project.

File: src/main/java/redis/clients/jedis/commands/JedisClusterCommands.java [hunk 4]
```diff
@@ -19,7 +19,10 @@
 import java.util.Map;
 import java.util.Set;
 
-public interface JedisClusterCommands {
+public interface JedisClusterCommands extends JedisBaseCommands {
+  
+  Long waitReplicas(final String key, final int replicas, final long timeout);
+  
   String set(String key, String value);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/JedisClusterCommands.java]: `JedisClusterCommands` shouldn't contain the methods from `JedisBaseCommands`.

**Comment by kamyuentse** [comment]
ping

**Comment by thattolleyguy** [comment]
What's the status of this issue? I'd be more than happy to help complete it. We upgraded to jedis 3.0 as part of upgrading to spring boot 2 but are now missing commands because the JedisClusterCommands and JedisCommands are not in sync.

**Comment by gkorland** [comment]
@thattolleyguy can you please be more specific about the missing commands? these are probably commands that are not supported by the cluster API

**Comment by thattolleyguy** [comment]
@gkorland These are the methods missing from `JedisCluster`:

- `hincrByFloat(String key, String field, double value)`
- `move(String key, int dbIndex)`
- `bitpos(String key, boolean value)`
- `bitpos(String key, boolean value, BitPosParams params)`
- `hscan(String key, String cursor, ScanParams params)`
- `sscan(String key, String cursor, ScanParams params)`
- `zscan(String key, String cursor, ScanParams params)`

All of them except `move` and `bitpos` are supported in the `BinaryJedisCluster` so as far as I can tell they should be supported in the `JedisCluster` as well. Maybe this is a separate issue but I thought it was related as they seem to be common commands between `Jedis` and `JedisCluster`.

Should the `move` and `bitpos` commands be supported in `JedisCluster`? I'm pretty new to redis so I don't know if they do or don't work in a cluster.

**Comment by sazzad16** [comment]
Resolved by #2533

**Pull request closed without merging**